src/python/pants/backend/docker/target_types.py

```diff
@@ -30,7 +30,11 @@ class DockerBuildArgsField(StringSequenceField):
 class DockerImageSourceField(SingleSourceField):
     default = "Dockerfile"
     required = False
-    help = "The Dockerfile to use when building the Docker image."
+    help = (
+        "The Dockerfile to use when building the Docker image.\n\n"
+        "This should either be a filename for a Dockerfile in your project workspace, or the "
+        "address to a `dockerfile` target."
+    )
```

Comment:
This is now no longer a breaking change!

However, it may be a bit controversial to support using a target address in a `source` field..?
I felt it was motivated in this case, as it will be the feature toggle for where to look for the `Dockerfile`.

When using a address for a `dockerfile` target, the dependency is injected, so it will show up in introspection etc.

Reply:
Example
```python
docker_image(source=":inline-dockerfile")
dockerfile(name="inline-dockerfile", instructions=[...])
```

Reply:
Hmm... yeah, this seems to not work well with `tailor` for instance... guess it's a bad idea, given that others may attempt to hydrate the sources as-is for any/all targets, and that would fail here...

Reply:
Pity, I thought it was a neat solution. Perhaps it would be OK if the field had some logic that by default presents itself as `None` in case it looks like an address... ?

Reply:
Ah, so I found a workaround, to return an empty `globs` in those cases. So, still controversial, but now it works with `tailor` and those who might want to hydrate this source field...

Reply:
Hm, I see now. If you use `dockerfile`, you must still use `docker_image`, right? Resulting in this cool but weird implementation

What about instead renaming the target to something like `synthetic_docker_image` and giving it roughly the same fields as `docker_image`? Meaning, all you need is `synthetic_docker_image`. It's the same except for having your `instructions` field rather than `source` field. 

The Target API is designed for reuse of fields like this, that multiple targets can have the same field. Your rules don't need to know the difference between the target types. So long as they use that `HydrateSourcesRequest(tgt.get(Sources), enable_codegen=True, for_sources_type=(DockerImageSourceField,))` snippet + implement the codegen rule, you should be good to go and everywhere else is agnostic to whether it's a `docker_image` or `synthetic_docker_image`.

(Another name could be `generated_docker_image`?)

Reply:
Yeah, you're right. The idea is that you still use `docker_image` regardless of where the Dockerfile comes from.
But it that so weird, though?
If I look at how protobuf is used in python code for instance, it seems similar to me (but the dependency is inferred rather than explicit).

The weirdest part here is putting the explicit dependency in the source field, but with several advantages. It signals that you may only have dependency to one `dockerfile` target, and it forces you to use either the target, or a file from the workspace, there's no room for both.

Of course, the same can be said about using two different targets for `docker_image`. But the naming of the alternative option here, with an inline or synthetic dockerfile, feels a bit off.
Do we have other cases where there are multiple targets that does semantically the same thing, but with different input sources?

Trying the various options out here for size..

```python
# OPT A
docker_image(
  source="Dockerfile",
  image_tags=["1.0"]
)

generated_docker_image(
  name="name-this",
  repository="my-repo/{name}",
  image_tags=["1.0"],
  instructions=[
    "FROM python:3.9",
    ...
  ]
)

## OR OPT B

docker_image(
  source=":my-dockerfile",
  image_tags=["2.0"]
)

dockerfile(
  name="my-dockerfile",
  instructions=[
    "FROM python:3.9",
    ...
  ]
)

## OR OPT C

docker_image(
  source="Dockerfile",
  image_tags=["1.0"]
)

docker_image(
  name="name-this",
  repository="my-repo/{name}",
  image_tags=["1.0"],
  instructions=[
    "FROM python:3.9",
    ...
  ]
)

docker_image(
  source="Dockerfile",
  image_tags=["1.0"]
  instructions=[
    "FROM python:3.9",
    ...
  ]
)
```

I think I prefer opt `C` if opt `B` is a no-go. That is, that we simply introduce a `instructions` field on the existing `docker_image`, and in case you specify both `source` __and__ `instructions`, you might as well merge them (that is OK, it will become a multistage build then, or if instructions doesn't have a `FROM`, it could be that it simply adds to what's already in the `Dockerfile` from the workspace).

Primarily because I can't think of a name for the `generated_docker_image` that makes sense to me.

Actually, the last option opens up a few interesting possibilities, when merging as mentioned above (that I don't have a direct need for, so is not a strong case, just interesting).

Reply:
Hmm... how about `dockerfile_image` for the synthetic version.. that's almost OK.. but I think it'll be slightly confusing, especially for new users.. (with whatever name we come up with, for opt `A`)

Reply:
@benjyw I'd be interested in your thoughts here as well.. :)

Reply:
> If I look at how protobuf is used in python code for instance, it seems similar to me (but the dependency is inferred rather than explicit).

Imo a difference there is that Protobuf is generating into a completely different language. And it might be generating into multiple languages. A more apples-to-apples comparison would be a Python target generating Python code that can be substituted into the `source` field for `python_source`.

--

I think Option A is reasonable. Although I think a downside is information overload, e.g. that there are two `help targets` entries that are nearly identical except for `source` vs. `instructions`.

--

I don't like Option B because of forcing you two have two targets, along with the `source` weirdness. The coupling smells wrong to me.

--

Option C looks reasonable to me! Although with a bias towards not merging until we have a use case. It's always possible to add that feature later, it's much harder to take it away or change the semantics once it's added.

I'd also suggest starting a new PR that does Option C, rather than updating this PR. Keep this one around as-is for posterity.

Reply:
I recommend using target validation to ensure that exactly one of those two fields is set:

https://github.com/pantsbuild/pants/blob/80bee0ad0977e83a2381530214fa3b362c805294/src/python/pants/engine/target.py#L590-L596

But hm, that is tricky that `source` has a default, which I think we really do want to keep as it's a great default the majority of the time. The Target API means that by the time you reach `Target.validate()`, you have no idea if the `Dockerfile` value is from the default vs. the user hardcoding it. So, two options:

1) Force the user to set `source=None`
2) Ignore `source` if `instructions` is set

Even though 1) is more boilerplate, I think that I prefer it a lot more. It's much more explicit, and it allows us to in the future allow both `source` and `instructions` at the same time. (Whereas it would be a breaking change to go from just ignoring `source` to now in the future using it.)

--

This conundrum does also give me some pause if Option A is better, as it avoids entirely any coupling, which is usually a smell with Target API modeling imo.

I'm still definitely open to Option C, although not 100% certain it's what I recommend

Reply:
I'm also starting to favour option A, based on that when I have source and instruction in the same target, codegen won't work. (Also your 1) above won't work, it would have to be `source=""`, which is ugly.)

Reply:
But to me, the information overload with duplicate information and very similar targets for `help targets` feels worse than having two targets with coupling (doesn't have to use the source field, could as well put it directly in `dependencies`).

Reply:
This is similar to `python_distribution`, where we may either generate a `setup.py` or use an existing one.

Reply:
I don't think we want source to be able to point to a target address, that is a recipe for confusion. This is what `dependencies` are for.

I think we want:

`docker_image()` as today, except that `source=` is now optional, and if you omit it, we look for a `dockerfile` target in the docker_image's dependencies, and if we don't find exactly 1 we error (and if we find any at all when there is also a `source=` we also error).

Reply:
This opens up future things like concatenating multiple `dockerfile` targets in your `dependencies`.

Reply:
> except that source= is now optional

A downside there is that we can't have a default of `Dockerfile` anymore, which is an excellent default. Feels overly verbose to have to set this:

```python
docker_image(source="Dockerfile")
```

Reply:
We can have a glob that matches Dockerfile or nothing?

Reply:
There will be a logged output at least about the glob not matching any files, I think. Can double check.

However, having "Dockerfile" as a default made sense, but it's not that much boilerplate saved with it, imo.

Reply:
yep, warning by default:
```
            default=FilesNotFoundBehavior.warn,
```

Reply:
OK, dropped the controversial bit where the dependency could be added via the `source` field. The `source` field is now optional, which is a breaking change (for a still experimental backend, so I think it should be OK without deprecations?)

> Feels overly verbose to have to set this (`source="Dockerfile"`)

I agree it was a good default, but I don't think it is in any way bad to provide this value either. It's only good as long as you have at most one `docker_image` target per BUILD file.. it also removes a bit of "magic" when being explicit about it.

I feel the alternative would be worse, having to provide `source=""` when using the `dockerfile` target to cancel the default value. If `source=None` would've worked, it would be slightly better, alas it does not.